Support Multi-User for @AutoOpen and @AsyncOpen

dianaafanador3 opened the pull request.

tgoyne commented on CHANGELOG.md:
```diff
@@ -6,6 +6,15 @@ x.y.z Release notes (yyyy-MM-dd)
 ### Fixed
 * <How to hit and notice issue? what was the impact?> ([#????](https://github.com/realm/realm-cocoa/issues/????), since v?.?.?)
 * None.
+* Add additional `observe` methods for Objects and RealmCollections which take a `PartialKeyPath` type key path parameter.
```
Bad changelog merge.

tgoyne commented on Realm.xcodeproj/project.pbxproj:
```diff
@@ -433,15 +433,11 @@
 		AC8846782686573B00DF4A65 /* ContentView.swift in Sources */ = {isa = PBXBuildFile; fileRef = AC8846772686573B00DF4A65 /* ContentView.swift */; };
 		AC8846B72687BC4100DF4A65 /* SwiftUIServerTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = AC8846B62687BC4100DF4A65 /* SwiftUIServerTests.swift */; };
 		AC88478626888CEE00DF4A65 /* SwiftUISyncTestHostUITests.swift in Sources */ = {isa = PBXBuildFile; fileRef = AC8846902686573D00DF4A65 /* SwiftUISyncTestHostUITests.swift */; };
-		AC8847A3268925F100DF4A65 /* RLMSyncTestCase.mm in Sources */ = {isa = PBXBuildFile; fileRef = 1AA5AE9B1D98A68E00ED8C27 /* RLMSyncTestCase.mm */; };
```
These changes seem suspicious.

dianaafanador3 replied:
I removed some files from the `SwiftUISyncTestHostUITests` target, because we are now doing all server setups on the same file , no need to inherit from `SwiftSyncTestCase`.

tgoyne commented on Realm/ObjectServerTests/SwiftUIServerTests.swift:
```diff
@@ -43,11 +45,26 @@ class SwiftUIServerTests: SwiftSyncTestCase {
 
         let ex = expectation(description: "download-realm-async-open")
-        asyncOpen(appId: appId, partitionValue: #function) { asyncOpenState in
+        asyncOpenForUser(user, appId: appId, partitionValue: #function) { asyncOpenState in
```
Idiomatic swift naming for this would be `asyncOpen(user: user, appId: appId`

tgoyne commented on RealmSwift/SwiftUI.swift:
```diff
@@ -675,52 +675,149 @@ extension EnvironmentValues {
+        app.objectWillChange.sink {
+            switch self.appState {
+            case .loggedIn(let user):
+                if let newUser = $0.currentUser,
```
This closure's doing enough stuff for it to be useful to name the argument.

tgoyne commented on RealmSwift/SwiftUI.swift:
```diff
@@ -847,38 +917,36 @@ public enum AsyncOpenState {
      - parameter configuration: The `Realm.Configuration` used when creating the Realm,
                  user's sync configuration for the given partition value will be set as the `syncConfiguration`,
-                 if empty the configuration is set to the `defaultConfiguration`
+                 if empty the user configuration will be used.
```
Isn't this a breaking change? If we previously were using settings from the default configuration and now aren't anyone who was relying on that will have problems.

jsflax replied:
This sounds like a bug in the original implementation– I believe we were always supposed to use the user's configuration.

dianaafanador3 replied:
yep as mentioned on the PR's description, this was a bug on the original implementation
`We are not setting defaultConfiguration as default if the user doesn't specified one in the initialisation of the property wrapper, this was causing the fileURL to be the one provided by the defaultConfiguration and never the one from the user configuration. This should only be valid if the user provides a configuration via the constructor, overriding the user configuration path.`

jsflax commented on Realm/Tests/SwiftUISyncTestHost/ContentView.swift:
```diff
@@ -98,43 +108,93 @@ struct LoginView: View {
     }
+
+    func logout() {
+        let appConfig = AppConfiguration(baseURL: "http://localhost:9090",
```
Should pull this config out somewhere.

jsflax commented on Realm/Tests/SwiftUISyncTestHostUITests/SwiftUISyncTestHostUITests.swift:
```diff
@@ -19,36 +19,68 @@
-                                                            in: .userDomainMask,
-                                                            appropriateFor: nil,
-                                                            create: false)
+                                                               in: .userDomainMask,
```
Bad indent.

jsflax commented on RealmSwift/SwiftUI.swift:
```diff
@@ -675,52 +675,149 @@ extension EnvironmentValues {
         }
     }
 }
+/**
```
Should have a new line before.

jsflax commented on RealmSwift/SwiftUI.swift:
```diff
@@ -675,52 +675,149 @@ extension EnvironmentValues {
+/**
+An enum representing the type of Async operation `@AsyncOpen` or `@AutoOpen`
+*/
+private enum AsyncOpenType {
```
Would prefer this be called `AsyncOpenKind`.

tgoyne commented on RealmSwift/Combine.swift:
```diff
@@ -715,6 +715,7 @@ public enum RealmPublishers {
             subscriber.receive(subscription: AsyncOpenSubscription(task: Realm.AsyncOpenTask(rlmTask: RLMRealm.asyncOpen(with: configuration.rlmConfiguration, callbackQueue: callbackQueue, callback: { rlmRealm, error in
                 if let realm = rlmRealm.flatMap(Realm.init) {
                     _ = subscriber.receive(realm)
+                    subscriber.receive(completion: .finished)
```
This looks like a bug fix worth mentioning in the changelog.

tgoyne commented on RealmSwift/SwiftUI.swift:
```diff
@@ -766,51 +766,149 @@ extension EnvironmentValues {
+An enum representing the type of Async operation `@AsyncOpen` or `@AutoOpen`
+*/
+private enum AsyncOpenKind {
+    /// @AsyncOpen
```
Since this is private it's fine to just leave it undocumented if there's nothing useful to say.

tgoyne reviewed (commented):
Code changes look good, but I would like to see the tests for this passing on CI. It looks to be some sort of problem in the setup/cleanup test code:

```
Oct 01 09:51:31 <unknown>:0: error: -[SwiftUISyncTestHostUITests.SwiftUISyncTestHostUITests testAsyncOpenAndLogout] : make_dir() failed: No such file or directory Path: /Users/realm/Library/Application Support/io.realm.TestHost.xctrunner/mongodb-realm/
Oct 01 09:51:31 Exception backtrace:
Oct 01 09:51:31 0   Realm                               0x0000000107b71bdb _ZN5realm4util12try_make_dirERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 491
Oct 01 09:51:31 1   Realm                               0x0000000107d6197f _ZN5realm15SyncFileManagerC2ERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEES9_ + 143
Oct 01 09:51:31 2   Realm                               0x0000000107d30e8b _ZN5realm11SyncManager9configureENSt3__110shared_ptrINS_3app3AppEEERKNS1_12basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEERKNS_16SyncClientConfigE + 299
Oct 01 09:51:31 3   Realm                               0x0000000107cfab92 _ZN5realm3app3App9configureERKNS_16SyncClientConfigE + 290
Oct 01 09:51:31 4   Realm                               0x0000000107cfa4d9 _ZN5realm3app3App14get_shared_appERKNS1_6ConfigERKNS_16SyncClientConfigE + 297
Oct 01 09:51:31 5   Realm                               0x0000000107825215 -[RLMApp initWithId:configuration:rootDirectory:] + 349
Oct 01 09:51:31 6   Realm                               0x000000010782561a +[RLMApp appWithId:configuration:rootDirectory:] + 158
```

This error indicates that the `io.realm.TestHost.xctrunner` directory doesn't exist.

dianaafanador3 commented:
@tgoyne I fixed the issue on the CI where `directory doesn't exist.`. But tests keeps failing because a login window interrupts the tests.
```
Oct 04 08:13:04     t =    65.02s         Found 1 interrupting element:
Oct 04 08:13:04     t =    65.02s             Other at {{0.0, 0.0}, {1024.0, 768.0}} from Application 'com.apple.loginwindow'
Oct 04 08:13:04     t =    65.02s         Invoking UI interruption monitors for Other at {{0.0, 0.0}, {1024.0, 768.0}} from Application 'com.apple.loginwindow'
```
I changed initialising `App` to use a rootDirectory, expecting the realm database data doesn't get encrypted and there is no need to access the keychain but it is not working, any suggestion I can test?. 
In my local, access to keychain is no longer required after I made the changes.
